consensus/src/data_manager.rs

```diff
@@ -0,0 +1,235 @@
+use aptos_logger::{debug, warn};
+use aptos_types::transaction::SignedTransaction;
+use arc_swap::ArcSwapOption;
+use consensus_types::common::DataStatus;
+use consensus_types::{
+    block::Block,
+    common::Payload,
+    proof_of_store::{LogicalTime, ProofOfStore},
+    request_response::WrapperCommand,
+};
+use executor_types::Error::BlockNotFound;
+use executor_types::*;
+use futures::channel::mpsc::Sender;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use tokio::sync::oneshot;
+
+/// Responsible to extract the transactions out of the payload and notify QuorumStore about commits.
+/// If QuorumStore is enabled, has to ask BatchReader for the transaction behind the proofs of availability in the payload.
+pub struct PayloadManager {
+    quorum_store_enabled: AtomicBool,
+    data_reader: ArcSwapOption<BatchReader>,
+    quorum_store_wrapper_tx: ArcSwapOption<Sender<WrapperCommand>>,
+}
+
+impl PayloadManager {
+    pub fn new() -> Self {
+        Self {
+            quorum_store_enabled: AtomicBool::new(false),
+            data_reader: ArcSwapOption::from(None),
+            quorum_store_wrapper_tx: ArcSwapOption::from(None),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn enable_quorum_store(&self) {
+        self.quorum_store_enabled.store(true, Ordering::Relaxed)
+    }
+
+    async fn request_transactions(
+        &self,
+        proofs: Vec<ProofOfStore>,
+        logical_time: LogicalTime,
+    ) -> Vec<oneshot::Receiver<Result<Vec<SignedTransaction>, executor_types::Error>>> {
+        let mut receivers = Vec::new();
+        for pos in proofs {
+            debug!(
+                "QSE: requesting pos {:?}, digest {}, time = {:?}",
+                pos,
+                pos.digest(),
+                logical_time
+            );
+            if logical_time <= pos.expiration() {
+                receivers.push(
+                    self.data_reader
+                        .load()
+                        .as_ref()
+                        .unwrap() //TODO: can this be None? Need to make sure we call new_epoch() first.
+                        .get_batch(pos)
+                        .await,
+                );
+            } else {
+                debug!("QS: skipped expired pos");
+            }
+        }
+        receivers
+    }
+
+    ///Pass commit information to BatchReader and QuorumStore wrapper for their internal cleanups.
+    pub async fn notify_commit(&self, logical_time: LogicalTime, payloads: Vec<Payload>) {
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            self.data_reader
+                .load()
+                .as_ref()
+                .unwrap()
+                .update_certified_round(logical_time)
+                .await;
+
+            let digests: Vec<HashValue> = payloads
+                .into_iter()
+                .flat_map(|payload| match payload {
+                    Payload::DirectMempool(_) => {
+                        unreachable!()
```

Comment:
at least add some messages, or we can just return vec![] too

Reply:
Replaced with:

warn!("InQuorumStore should be used");
Vec::new()